dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/EventDateValidationHook.java

```diff
@@ -0,0 +1,197 @@
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.program.ProgramInstance;
+import org.hisp.dhis.program.ProgramStageInstance;
+import org.hisp.dhis.trackedentity.TrackedEntityInstance;
+import org.hisp.dhis.tracker.bundle.TrackerBundle;
+import org.hisp.dhis.tracker.domain.Event;
+import org.hisp.dhis.tracker.preheat.PreheatHelper;
+import org.hisp.dhis.tracker.report.TrackerErrorCode;
+import org.hisp.dhis.tracker.report.TrackerErrorReport;
+import org.hisp.dhis.tracker.report.ValidationErrorReporter;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.util.DateUtils;
+import org.springframework.stereotype.Component;
+
+import java.util.Date;
+import java.util.List;
+
+import static org.hisp.dhis.tracker.report.ValidationErrorReporter.newReport;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@Component
+public class EventDateValidationHook
+    extends AbstractTrackerValidationHook
+{
+
+    @Override
+    public int getOrder()
+    {
+        return 302;
+    }
+
+    @Override
+    public List<TrackerErrorReport> validate( TrackerBundle bundle )
+    {
+        ValidationErrorReporter reporter = new ValidationErrorReporter( bundle, this.getClass() );
+        User actingUser = bundle.getPreheat().getUser();
+
+        for ( Event event : bundle.getEvents() )
+        {
+            reporter.increment( event );
+
+            ProgramStageInstance programStageInstance = PreheatHelper
+                .getProgramStageInstance( bundle, event.getEvent() );
+            ProgramInstance programInstance = PreheatHelper.getProgramInstance( bundle, event.getEnrollment() );
+            TrackedEntityInstance trackedEntityInstance = PreheatHelper
+                .getTrackedEntityInstance( bundle, event.getTrackedEntity() );
+            Program program = PreheatHelper.getProgram( bundle, event.getProgram() );
+
+            if ( EventStatus.ACTIVE == event.getStatus() && event.getEventDate() == null )
+            {
+                reporter.addError( newReport( TrackerErrorCode.E1031 )
+                    .addArg( event ) );
+                continue;
+            }
+
+            if ( program == null )
+            {
+                continue;
+            }
+
+            programInstance = getProgramInstance( actingUser, programInstance, trackedEntityInstance, program );
+            program = programInstance.getProgram();
+
+            validateExpiryDays( reporter, event, program, programStageInstance );
+            validateDates( reporter, event );
+        }
+
+        return reporter.getReportList();
+    }
+
+    private void validateExpiryDays( ValidationErrorReporter errorReporter,
+        Event event,
+        Program program,
+        ProgramStageInstance programStageInstance )
+    {
+//        if ( importOptions == null || importOptions.getUser() == null ||
+//            importOptions.getUser().isAuthorized( Authorities.F_EDIT_EXPIRED.getAuthority() ) )
+//        {
+//            return;
+//        }
```

Comment:
Not sure why this is turned off, otherwise it is an important feature.

The intention is to ignore validating expiry dates under such conditions ... this is especially important for example when importing bulk events.

Reply:
Added check for F_EDIT_EXPIRED. In the new importer, I hope we can stop making null check on the user to allow for such special cases.